internals-js/src/values.ts

```diff
@@ -61,18 +75,26 @@ export function valueToString(v: any, expectedType?: InputType): string {
 
   if (Array.isArray(v)) {
     let elementsType: InputType | undefined = undefined;
-    if (expectedType) {
-      if (!isListType(expectedType)) {
-        throw buildError(`Invalid list value for non-list type ${expectedType}`);
-      }
+    // If the expected type is not a list, we've been given an invalid type. We don't want this
+    // method to fail though, so we just ignore the provided type from that point one (passing
+    // `undefined` to the recursion).
+    if (expectedType && isListType(expectedType)) {
       elementsType = expectedType.ofType;
     }
     return '[' + v.map(e => valueToString(e, elementsType)).join(', ') + ']';
   }
 
+  // We know the value is not a list/array. But if the type is a list, we still want to print
+  // the value correctly, at least as long as it's a valid value for the element type, since
+  // list input coercions may allow this.
+  if (expectedType && isListType(expectedType)) {
+    expectedType = expectedType.ofType;
```

Comment:
Note you need to unwrap multiple times for nesting (e.g. `[[[Foo!]!]!]`), so I think this might need to be:

```suggestion
    expectedType = expectedType.ofType;
    return valueToString(v, expectedType);
```

Reply:
Oh yeah, good point. Going fix and add a test.